network/src/peers/peer_book.rs

```diff
@@ -14,503 +14,263 @@
+                if peer.judge() {
+                    // dont reconnect to bad peers
+                    return Ok(None);
+                }
             }
-            // Case 2 - A peer that was previously not known.
-            None => PeerInfo::new(listener),
-        };
-
-        // Remove the peer's address from the list of connecting peers.
-        if self.connecting_peers.write().remove(&address) {
-            metrics::decrement_gauge!(stats::CONNECTIONS_CONNECTING, 1.0);
+            let peer = if let Some(peer) = self.take_disconnected_peer(address).await {
+                peer
+            } else {
+                Peer::new(address, node.config.bootnodes().contains(&address))
+            };
+            self.pending_connections.fetch_add(1, Ordering::SeqCst);
+            peer.connect(node, self.peer_events.clone());
+            Ok(None)
         }
+    }
 
-        // Update the peer info to connected.
-        peer_info.set_connected();
-
-        // Add the address into the connected peers.
-        if self.connected_peers.write().insert(listener, peer_info).is_none() {
-            metrics::increment_gauge!(stats::CONNECTIONS_CONNECTED, 1.0);
-        } else {
-            error!("{} is already a connected peer!", listener);
+    /// concurrently iterates over peers
+    async fn for_each_peer<F: Future<Output = ()>, FN: Fn(PeerHandle) -> F>(&self, func: FN) {
+        let mut futures = vec![];
+        for (_, peer) in self.connected_peers.inner().iter() {
+            futures.push(func(peer.clone()));
         }
+        futures::future::join_all(futures).await;
     }
 
-    ///
-    /// Removes the given address from the connecting and connected peers in this `PeerBook`,
-    /// and adds the given address to the disconnected peers in this `PeerBook`.
-    ///
-    pub fn set_disconnected(&self, address: SocketAddr) -> bool {
-        // Case 1 - The given address is a connecting peer, attempt to disconnect.
-        if self.connecting_peers.write().remove(&address) {
-            metrics::decrement_gauge!(stats::CONNECTIONS_CONNECTING, 1.0);
-            return false;
+    /// concurrently iterates over peers
+    async fn map_each_peer<O: Send + Sync, F: Future<Output = Option<O>>, FN: Fn(PeerHandle) -> F>(
+        &self,
+        func: FN,
+    ) -> Vec<O> {
+        let mut futures = Vec::with_capacity(self.connected_peers.len());
+        for (_, peer) in self.connected_peers.inner().iter() {
+            futures.push(func(peer.clone()));
         }
+        futures::future::join_all(futures).await.into_iter().flatten().collect()
+    }
 
-        // Case 2 - The given address is a connected peer, attempt to disconnect.
-        if let Some(mut peer_info) = self.connected_peers.write().remove(&address) {
-            // Update the peer info to disconnected.
-            peer_info.set_disconnected();
-
-            metrics::decrement_gauge!(stats::CONNECTIONS_CONNECTED, 1.0);
+    pub async fn judgement(&self) {
```

Comment:
this name is not informative enough; it could be e.g. `check_quality_of_peers` or `evaluate_peers`